### ISSUE TITLE
RuntimeWarning: invalid value encountered in cast in pose_estimation_pytorch/data/postprocessor.py

### ISSUE THREAD

**Issue by emilyzys** [DeepLabCut/DeepLabCut#3376]
Hello!

I am trying to train a model on rat cylinder test videos, and the eventual goal is to score forepaw cylinder wall contacts automatically. The videos are recorded with a mirror, so each frame contains both the real rat and its mirror reflection. Here is a screenshot of one of the videos.

<img width="2416" height="1368" alt="Image" src="https://github.com/user-attachments/assets/df74be68-dd08-4a05-93f0-dbee0f2c5887" />

I initially tried two different project designs with small number of frames(69) just as a preliminary experiment to compare the two approaches:

1.  Label the following body parts as a single animal:
real_nose
real_left_forepaw
real_right_forepaw
real_tailbase
mirror_nose
mirror_left_forepaw
mirror_right_forepaw
mirror_tailbase

2. Treat the real rat and mirror rat as two individuals and used multi-animal training:
Individual 1 = real rat
Individual 2 = mirror rat

Body parts:
nose
left_forepaw
right_forepaw
tailbase

After training the two models with the exact same frames and number of epochs, the multi-animal model perfromed much better, with train and test rmse about 14, whereas for the single animal model the rmses are around 200. However, the multi-animal model encountered problems at analyze_videos step: when running the code shell in colab, I consistently get the warning /usr/local/lib/python3.12/dist-packages/deeplabcut/pose_estimation_pytorch/data/postprocessor.py:507:
RuntimeWarning: invalid value encountered in cast for all my videos.
As a result, there is a large numbers of blanks in the output CSV. And correspond to that, in the created labeled videos, many frames contain predictions for only the real rat or only the mirror rat and some frames contain neither. Lowering the pcutoff to 0, disable auto_track, changing tracking method or changing batch size did not help.

Is this problem is likely due to:

Too few/diverse training frames, and would likely improve with 200–500+ labeled frames?
A limitation of the multi-animal assembly/tracking pipeline when the two "individuals" are actually a real animal and its mirror reflection?
A DLC 3.0.0 issue?

Would you recommend continuing with the multi-animal approach and collecting many more training frames, or switching back to the single-animal approach and focusing on improving that model instead?

Any advice would be greatly appreciated.

Thanks,
Emily

**Comment by YasinKaryagdi** [comment]
Heyhi! 
Quick disclaimer: not an expert on DLC but I've some thoughts that might help, especially as no one has yet replied:
1. I would not proceed with this single animal setup. I'd argue that a nose is a nose, and that there might not be any real distinction between a mirror nose and a real nose. So to me this seems like "there are two animals". So i'd proceed with the multi animal case. Another alternative would be to maybe cut up the video in two parts by splitting it in the middle and change the skeleton, that way you'd have a single animal case, though I'd not recommend this approach. As it might take some more effort gluing everything together. Though you'd probably just store 1 video as {curr_vidname} + "_real" and the other as {curr_vidname} + "_mirror". But again, multi animal approach might be more productive.
2. What kind of model are you running for the multi animal case? Are you using pretrained weights? Are you using a topdown or bottom up approach? My guess is that the jar is fixed and therefore there is most likely little/none overlap between the real and mirror mouse, therefore I believe a top down model would preform better, though you could try both. My concern is that it's actually unclear what your whole setup is, so it's hard to pinpoint what the issue might be. What batch size did you try? How many epochs did you run it for? Were you consistent in annotations? And if so was the samples diverse enough to capture all the various rotations? Also do you notice any clear pattern in the cases where there are no predictions? Are these represented in your samples?

Wishing you much luck with the project, it seems interesting!
Kind regards,
Yasin

**Comment by deruyter92** [comment]
Hi @emilyzys, 


**Regarding the warning**:  I think you identified an issue with our code here. Probably, when the output of the estimation model contains partial assemblies (NaN values), the identity postprocessor doesn't handle this well and throws a warning (and identity assignment may be silently wrong). A few follow-up questions to confirm this diagnosis: 
is `identity: true` in your project configuration? what `net_type` did you use? Does the warning disappear when you set `identity: false`?

I've addressed the issue in #3378. You can try to install it using the following command:
```
pip install --upgrade --force-reinstall git+https://github.com/DeepLabCut/DeepLabCut.git@jaap/fix_postprocessor_nan_pafs
```


**Regarding your project:**
Unfortunately, solving the aforementioned issue is probably not going to improve your results. (It is a non-fatal warning that likely does not influence your analysis results). 

I agree with @YasinKaryagdi that it is probably better to proceed with a multi-animal workflow. What you maybe could try is 
- make sure that your labels are as complete as possible
- train with a top-down model instead of bottom up

That said, I'd prefer to continue the discussion on your [forum post](https://forum.image.sc/t/single-animal-vs-multi-animal-approach-for-tracking-videos-with-mirror-reflection/121531), which is indeed a great place to ask for tips about your project! We try to reserve the github issues sections for problems that are directly related to the code. 

Thanks for raising the issue, let me know if you tried out our fix and if the warning persists.

**Comment by emilyzys** [comment]
Hi,

Thank you both very much for the suggestions!

I set identity to false and installed the update. The warning disappeared, but as you predicted, the analysis result still have the same problem of missing labels. I will continue the discussion on my forum post.

Emily
